### DISCUSSION TITLE
CHAPTER 5 - QnA

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#53] [CHAPTER 5 스트림 활용 - Q&A]
p. 175
> 컬렉션으로 스트림을 만드는 stream 메서드를 parallelStream 로 바꾸는 것만으로도 별다른 노력없이 병렬성을 얻을 수 있다.  ~~~ 심지어 이런 계산을 병렬로 실행할 수 있다. 
하지만 이들은 각각 다양한 연산을 수행한다. 따라서 각각의 연산은 내부적인 상태를 고려해야 한다.

Q1. Stream API 에서 parallel() 과 parallelStream() 으로 병렬 실행을 할 수 있는데, 언제 병렬 실행하는 것이 좋을까? 

Q2. 책에서 
`map`, `filter` 같은 상태를 갖지 않는 연산 (stateless operation)과 
`reduce`, `sum`, `max` 같은 상태 크기가 한정적인 연산(bounded stateful operation) 과
`sorted` 과 `distinct` 와 같은 상태 크기가 정해지지 않은 연산 (unbounded stateful operation) 을 설명한다.

- 상태를 가지는 연산을 작성할 때 주의할 점이 있을 것 같은데, 어떤 점이 있을까?
- `상태 크기가 정해지지 않은 연산 (unbounded stateful operation)` 은 무한 스트림을 사용할 때 문제가 생길 수 있다. 
    - 그렇다면 무한 스트림에서 `sorted` 나  `distinct`같은 연산을 사용할 경우, 문제를 피할 수 있는 방법이 있을까?

---

p 182.

> 5.7.1 기본형 특화 스트림
> 
- 자바 8에서는 세가지 기본형 특화 스트림을 제공한다. 스트림 API는 박싱 비용을 피할 수 있도록 int 요소에 특화된 `IntStream`, double 요소에 특화된 `DoubleStream`, long 요소에 특화된 `LongStream` 을 제공한다.

Q3. 왜 기본형 특화 스트림은 3가지만 제공할까?
(나머지 (e.g. short, char) 형들에 대해서는 왜 제공하지 않을까)

**Comment by JoisFe** [comment]
A1) 연산 순서가 의미가 있지 않고, 병렬작업이 결과에 영향을 미치는 상황이 아니며 빠른 결과를 내야한다면 parallelStream()을 사용하면 좋지 않을까요?
A2) 무한한 스트림에서 정렬, 혹은 중복 제거를 할 수 없기 때문 아닐까요? 끝이 없는데 무한한 위치에 있는 값과 첫번째 값을 비교한다는게 힘들 것 같아요

**Comment by BvrPark** [comment]
A1) 윗분의 의견과 같습니다. `parallelStream()`이 스트림을 실행할 때, 병렬 실행을 설정하기 때문에 기존 스트림을 병렬 실행으로 변환하는 `parallel()`보다는 더 빠르지 않을까요?

A2) 스트림이 저장이 아닌 단순히 데이터를 읽는 것이면 무한한 스트림도 데이터를 나눠서 읽을 수 있지 않을까요...?
만약 나눠서 읽을 수 있다면 그때, `sorted` , `distinct`와 같은 연산을 사용할 수 있을 것 같습니다.
